### PR TITLE
feat: [CDS-70206]: Implement fetch helm chart metadata

### PR DESCRIPTION
**Pull request by abosii-harness**
## Describe your changes
Implement fetch helm chart metadata task. The task will allow to fetch the chart.yaml from the target repository and retrieve the content. 

 The supported repository types are: Git, HTTP, OCI, S3 and GCS. 

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/49390)

### CODE REVIEW

**Comment by abosii-harness** [comment]
trigger checkstylecodeformat, commonchecks

**Comment by abosii-harness** [comment]
trigger checkstylecodeformat, commonchecks

**Comment by abosii-harness** [comment]
trigger checkstylecodeformat, commonchecks

**Comment by abosii-harness** [comment]
trigger checkstylecodeformat, commonchecks

**Comment by abosii-harness** [comment]
trigger uts

**Comment by abosii-harness** [comment]
trigger checkstylecodeformat, commonchecks

**Comment by abosii-harness** [comment]
trigger uts

**Review by alebuh — APPROVED** [review]
> Reviewed 17 of 17 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49390)* status: 17 file(s) to review. 1 of 2 approvals obtained. Waiting on team(s): k8s-team.

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/base/HelmChartManifestTaskService.java [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+    } else {
+      helmTaskHelperBase.downloadHelmChart(
+          manifestConfig, timeoutInMillis, new NoopExecutionCallback(), destinationDirectory);
+    }
+  }
+
+  private String getChartYamlPath(HelmChartManifestDelegateConfig manifestConfig, String basePath) {
+    if (isNotEmpty(manifestConfig.getSubChartPath())) {
+      return Paths.get(basePath, manifestConfig.getSubChartPath(), CHARTS_YAML_KEY).toString();
+    } else {
```
Comment (abhinavsingh1196) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/base/HelmChartManifestTaskService.java:193]: nit: redundant else, we can just return directly

**Review by abhinavsingh1196 — APPROVED** [review]

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/base/HelmChartManifestTaskService.java [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+@Slf4j
+@Singleton
+@OwnedBy(HarnessTeam.CDP)
+public class HelmChartManifestTaskService {
+  @Inject private HelmTaskHelperBase helmTaskHelperBase;
+
+  @Inject private GitFetchTaskHelper gitFetchTaskHelper;
+
+  private final Cache<HelmChartKey, HelmChartManifest> cache =
+      CacheBuilder.newBuilder().expireAfterAccess(Duration.ofHours(1)).expireAfterWrite(Duration.ofHours(2)).build();
```
Comment (GoelPratyush99) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/base/HelmChartManifestTaskService.java:74]: Can we add the number of hours as Static final variables

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/base/HelmChartManifestTaskService.java [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+
+  public HelmChartManifest fetchHelmChartManifest(FetchHelmChartManifestRequest request) throws Exception {
+    HelmChartMetadata metadata = createMetadata(request.getManifestDelegateConfig());
+    Optional<HelmChartKey> helmChartKey = metadata.toHelmChartKey();
+    if (helmChartKey.isPresent()) {
+      return cache.get(helmChartKey.get(), () -> fetchAndReadHelmChartYaml(request, metadata));
+    } else {
+      return fetchAndReadHelmChartYaml(request, metadata);
```
Comment (GoelPratyush99) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/base/HelmChartManifestTaskService.java:82]: nit: redundant else, we can return directly.

**Comment by abosii-harness** [comment]
trigger checkstylecodeformat, commonchecks

**Comment by abosii-harness** [comment]
trigger uts

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/base/HelmChartManifestTaskService.java [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+      CacheBuilder.newBuilder()
+          .expireAfterAccess(Duration.ofHours(CHART_CACHE_EXPIRE_AFTER_ACCESS_H))
+          .expireAfterWrite(Duration.ofHours(CHART_CACHE_EXPIRE_AFTER_WRITE_H))
+          .build();
+
+  public HelmChartManifest fetchHelmChartManifest(FetchHelmChartManifestRequest request) throws Exception {
+    HelmChartMetadata metadata = createMetadata(request.getManifestDelegateConfig());
+    Optional<HelmChartKey> helmChartKey = metadata.toHelmChartKey();
+    if (helmChartKey.isPresent()) {
+      return cache.get(helmChartKey.get(), () -> fetchAndReadHelmChartYaml(request, metadata));
```
Comment (acasian-harness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/base/HelmChartManifestTaskService.java:86]: trying to understand the logic, if the key is present in cache : retrieve it, else fetch & store in cache?

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/helm/request/HelmFetchChartManifestTaskParameters.java [hunk 5]
```diff
@@ -27,7 +27,9 @@
 @Builder
 @OwnedBy(HarnessTeam.CDP)
 public class HelmFetchChartManifestTaskParameters implements TaskParameters, ExecutionCapabilityDemander {
+  String accountId;
   @Expression(ALLOW_SECRETS) HelmChartManifestDelegateConfig helmChartConfig;
+  long timeoutInMillis;
```
Comment (acasian-harness) [inline 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/helm/request/HelmFetchChartManifestTaskParameters.java:32]: where is this being used?

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/HelmFetchChartManifestTask.java [hunk 6]
```diff
@@ -36,7 +58,41 @@ public DelegateResponseData run(Object[] parameters) {
...
+              HelmFetchChartManifestTaskParameters.class.getSimpleName())));
+    }
+    final HelmFetchChartManifestTaskParameters fetchTaskParameters = (HelmFetchChartManifestTaskParameters) parameters;
+
+    final String manifestDirectory = Paths.get(RESOURCE_DIR_BASE, convertBase64UuidToCanonicalForm(generateUuid()))
+                                         .normalize()
+                                         .toAbsolutePath()
+                                         .toString();
+
+    try (AutoCloseableWorkingDirectory _directory = new AutoCloseableWorkingDirectory(manifestDirectory, 10)) {
```
[the remark this answers is not in the document]
Comment (abosii-harness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/HelmFetchChartManifestTask.java:73]: I checked, I guess is not a java convention. Guess I have this practice from other language, I'll rename it

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/base/HelmChartManifestTaskService.java (hunk 4 above)
Comment (abosii-harness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/base/HelmChartManifestTaskService.java:86]: yes, this is the right behavior

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/helm/request/HelmFetchChartManifestTaskParameters.java (hunk 5 above)
Comment (abosii-harness) [reply 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/helm/request/HelmFetchChartManifestTaskParameters.java:32]: Is required for fetch helm repos methods

**Comment by abosii-harness** [comment]
trigger checkstylecodeformat, commonchecks

**Comment by abosii-harness** [comment]
trigger uts

**Comment by abosii-harness** [comment]
trigger ut0

**Comment by abosii-harness** [comment]
trigger ut0

**Comment by abosii-harness** [comment]
trigger checkstylecodeformat, commonchecks

**Comment by abosii-harness** [comment]
trigger uts

**Comment by GoelPratyush99** [comment]
you should rebase your branch. The checks should pass now.

**Comment by abosii-harness** [comment]
trigger checkstylecodeformat, commonchecks

**Comment by abosii-harness** [comment]
trigger uts

**Review by TarunUba1 — APPROVED** [review]

**Comment by TarunUba1** [comment]
Can you also add Testing regarding this PR

**Review by GoelPratyush99 — APPROVED** [review]

**Comment by abosii-harness** [comment]
trigger messagecheck

**Comment by abosii-harness** [comment]
trigger messagecheck

**Pull request merged**